src/rnnlm/rnnlm-simple-looped.cc

```diff
@@ -0,0 +1,200 @@
+  CreateLoopedComputationRequestSimple(rnnlm, frames_per_chunk,
+                                       frame_subsampling_factor,
+                                       ivector_period,
+                                       0, // extra_left_context_initial == 0
+                                       extra_right_context,
+                                       num_sequences,
+                                       &request1, &request2, &request3);
+
+  CompileLooped(rnnlm, opts.optimize_config, request1, request2,
+                request3, &computation);
+  computation.ComputeCudaIndexes();
+  if (GetVerboseLevel() >= 3) {
+    KALDI_VLOG(3) << "Computation is:";
+    computation.Print(std::cerr, rnnlm);
+  }
+}
+
+RnnlmSimpleLooped::RnnlmSimpleLooped(
+    const RnnlmSimpleLoopedInfo &info) :
+    info_(info),
+    computer_(info_.opts.compute_config, info_.computation,
+              info_.rnnlm, NULL),  // NULL is 'nnet_to_update'
+    // since everytime we provide one chunk to the object, the size of
+    // feats_ == frames_per_chunk
+    feats_(info_.frames_per_chunk,
+           info_.word_embedding_mat.NumRows()),
+    current_log_post_offset_(-1)
+{}
+
+RnnlmSimpleLooped::RnnlmSimpleLooped(const RnnlmSimpleLooped &other):
+  info_(other.info_), computer_(other.computer_), feats_(other.feats_),
+  current_nnet_output_(other.current_nnet_output_),
+  current_log_post_offset_(other.current_log_post_offset_)
+{}
+
+void RnnlmSimpleLooped::TakeFeatures(
+    const std::vector<int32> &word_indexes) {
+  KALDI_ASSERT(word_indexes.size() == feats_.NumRows());
+  std::vector<std::vector<std::pair<MatrixIndexT, BaseFloat> > >
+      pairs(word_indexes.size());
+  for (int32 i = 0; i < word_indexes.size(); i++) {
+    std::pair<MatrixIndexT, BaseFloat> one_hot_index(word_indexes[i], 1.0);
+    std::vector<std::pair<MatrixIndexT, BaseFloat> > row(1, one_hot_index);
+    pairs[i] = row;
+  }
+  SparseMatrix<BaseFloat> feats_temp(feats_.NumCols(), pairs);
+  feats_.Swap(&feats_temp);
+  // resets offset so that AdvanceChunk() would be called in GetOutput() and
+  // GetNnetOutputForFrame() after taking new features
+  current_log_post_offset_ = -1;
+}
+
+//void RnnlmSimpleLooped::GetNnetOutputForFrame(
+//    int32 frame, VectorBase<BaseFloat> *output) {
+//  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+//  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+//    AdvanceChunk();
+//  output->CopyFromVec(current_nnet_output_.Row(frame -
+//                                               current_log_post_offset_));
+//}
+
+BaseFloat RnnlmSimpleLooped::LogProbOfWord(int32 word_index,
+                               const CuVectorBase<BaseFloat> &hidden) const {
+  const CuMatrix<BaseFloat> &word_embedding_mat = info_.word_embedding_mat;
+  BaseFloat log_prob;
+
+  if (info_.opts.force_normalize) {
+    CuVector<BaseFloat> log_probs(word_embedding_mat.NumRows());
```

Comment:
if you are going to normalize, you should do this just once when you create the state, not each time you process an output arc.